### PR TITLE
Implement ecrecover precompile

### LINKED ISSUE #436: [EVM] re-enable ecrecover as a native function

### PR DESCRIPTION
**Pull request by silasdavis**
fixes #436

Main tangential thing fixed here was rlp encoding and decoding that was not working for long strings of slices. 

I also ended up refactoring away our use of non pointer public keys which aligns better with all the libraries we use. 

The eth test now generated its own raw tx for `TestEthSendRawTransaction` and I have added `crypto/signature` code to handle serialising between compact sigs in btcec and ethereum format

### CODE REVIEW

**Comment by silasdavis** [comment]
This Yak is being shorn:

- Hard-coded values in eth test break because our secp256k1 derivations was wrong (and fixed in first commit)
- Try to use RLP to generate raw eth transaction
- RLP is broken in various ways, particularly decoding
- Fix RLP
- Make public keys pointers on a whim
- Refactor

Nearly there though

File: encoding/rlp/rlp.go [hunk 1]
```diff
@@ -131,211 +219,119 @@ func encode(input interface{}) ([]byte, error) {
...
-	fields [][]byte
-}
-
-func (f *fields) add(element []byte) {
-	f.fields = append(f.fields, element)
-}
-
-func decode(in []byte, out *fields) error {
+// Split into RLP fields by reading length prefixes and consuming chunks
+func decode(in []byte) ([][]byte, error) {
```
Comment (silasdavis) [inline encoding/rlp/rlp.go:241]: This seemed to biggest place that was non-compliant

File: rpc/eth_test.go [hunk 2]
```diff
@@ -107,14 +110,34 @@ func TestWeb3Service(t *testing.T) {
...
 
+		sender := genesisAccounts[1]
 		receivee := genesisAccounts[2].GetPublicKey().GetAddress()
 		acc, err := kern.State.GetAccount(receivee)
 		require.NoError(t, err)
 		before := acc.GetBalance()
 
 		t.Run("EthSendRawTransaction", func(t *testing.T) {
-			// see: https://github.com/ethereumjs/ethereumjs-tx/blob/master/examples/transactions.ts#L9
-			raw := `0xf867808082520894f97798df751deb4b6e39d4cf998ee7cd4dcb9acc880de0b6b3a76400008025a0f0d2396973296cd6a71141c974d4a851f5eae8f08a8fba2dc36a0fef9bd6440ca0171995aa750d3f9f8e4d0eac93ff67634274f3c5acf422723f49ff09a6885422`
```
Comment (silasdavis) [inline rpc/eth_test.go:117]: This was the main thing I needed to fix since our secp generated addresses had changed I didn't know how to recreate this

**Comment by silasdavis** [comment]
Will wait on review from @seanyoung on this

File: crypto/signature.go [hunk 3]
```diff
@@ -2,12 +2,99 @@ package crypto
...
+	V big.Int `json:"v"`
+	R big.Int `json:"r"`
+	S big.Int `json:"s"`
+}
+
+// Returns either 0 or 1 for the underlying parity of the public key solution
+func (s *Secp256k1Signature) Parity() uint {
+	// odd  V => parity = 0
+	// even V => parity = 1
+	return s.V.Bit(0) ^ 0x01
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply crypto/signature.go:33]: int popcount = 0;
for b in s.V.Bytes() {
popcount += bits.OnesCount8(b);
}
return popcount & 1;

**Review by seanyoung — COMMENTED** [review]
Parity looks wrong, otherwise looks good

File: crypto/signature.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (silasdavis) [reply crypto/signature.go:33]: I had the other parity, the one you mention, in mind: https://en.wikipedia.org/wiki/Parity_(mathematics) :)

My parity is older than your parity :-p

However, I'll admit in this context it is potentially confusing. I'd change it to `IsEven` but it still depends on a bit convention for truthiness which varies for Ethereum for example. `Parity` kind of captures this - it's a function that assigns boolean based on the.

I'll change it to `RecoveryIndex`

[the remark this answers is not in the document]
Comment (seanyoung) [reply crypto/signature.go:33]: Maybe mention this in the comments.

**Pull request merged**